src/main/scala/com/ing/wbaa/gargoyle/proxy/provider/LineageProviderAtlas.scala

```diff
@@ -0,0 +1,143 @@
+      }
+      processGuid <- method match {
+        case access if access == "read" =>
+          postData(
+            processEntities(serverGuid, bucketGuid, fileGuid, userSTS, method, bucketGuid, "Bucket", fileGuid, "DataFile", timestamp).toJson)
+        case access if access == "write" =>
+          postData(
+            processEntities(serverGuid, bucketGuid, fileGuid, userSTS, method, fileGuid, "DataFile", bucketGuid, "Bucket", timestamp).toJson)
+      }
+    } yield Some(LineagePostGuidResponse(serverGuid, bucketGuid, fileGuid, processGuid))
+  }
+
+  // file entity delete
+  // for now it is just deleting file entity and no related objects like eg. aws_cli_script, which uploaded or downloaded
+  // file to bucket. Once we delete aws_cli_script object we will lose track of whats has been deleted
+  // We need to come up with process of tracking file delete
+  private def deleteEntities(typeName: String, entityName: String): Future[Option[LineagePostGuidResponse]] = {
+    for {
+      entityGuid <- getEntityGUID(typeName, entityName)
+      _ <- deleteEntity(entityGuid)
+
+    } yield (Some(LineagePostGuidResponse("", "", entityGuid, "")))
+  }
+
+  def createLineageFromRequest(httpRequest: HttpRequest, userSTS: User): Future[Option[LineagePostGuidResponse]] = {
+
+    val timestamp = System.currentTimeMillis()
+    // createLineageFromRequest is used in RequestHandlerS3 where there is no notion of S3Request. It seems easier to
+    // repeat httpRequest parsing here for needed values
+    val host = httpRequest.uri.authority.host.address()
+    val path = httpRequest.uri.path
+    val bucket = path.toString.split("/").toList.lift(1).getOrElse("notDef")
+    val bucketObject = s"${path.toString.split("/").toList.lift(2).getOrElse("emptyObject")}"
+    val method = httpRequest.method
+    val contentType = httpRequest.entity.contentType
+    val userName = userSTS.userName.value
+
+    if (bucket != "notDef" && bucketObject != "emptyObject") {
+      logger.debug(s"Creating lineage for request ${method.value} file ${bucketObject} in ${bucket} at ${timestamp}")
+      method match {
+        case HttpMethods.GET =>
+          logger.debug(s"Creating Read lineage for request to ${method.value} file ${bucketObject} to ${bucket} at ${timestamp}")
+          postEnities(userName, host, bucket, bucketObject, "read", contentType, timestamp)
+        case HttpMethods.POST | HttpMethods.PUT =>
+          logger.debug(s"Creating Write lineage for request to ${method.value} file ${bucketObject} to ${bucket} at ${timestamp}")
+          postEnities(userName, host, bucket, bucketObject, "write", contentType, timestamp)
+        case HttpMethods.DELETE =>
+          logger.debug(s"Creating Delete lineage for request to ${method} file ${bucketObject} to ${bucket} at ${timestamp}")
+          deleteEntities("DataFile", bucketObject)
+        case _ => Future(None)
+      }
+    } else {
+      Future(None)
```

Comment:
instead of a Future[Option[...]] we could also return just a Future[...]. Here you could let the future fail since in this case it is actually a failed attempt to write lineage

Reply:
yes, changed to feature failed